Close service when prepared query can't be parsed

jeremyrsmith opened the pull request:
This fixes an issue wherein the service isn't closed when the parse
phase of a prepared statement fails.  This causes that connection to
hang.  Instead, the parse failure should cause the service to be closed
(just as the successful completion of the query closes the service, and
the failure at any other point during the process does the same).

Commit:
Close service when prepared query can't be parsed

This fixes an issue wherein the service isn't closed when the parse
phase of a prepared statement fails.  This causes that connection to
hang.  Instead, the parse failure should cause the service to be closed
(just as the successful completion of the query closes the service, and
the failure at any other point during the process does the same).

vkostyukov commented on src/main/scala/com/twitter/finagle/postgres/Client.scala:
```diff
@@ -81,6 +81,8 @@ class Client(factory: ServiceFactory[PgRequest, PgResponse], id:String) {
         parse(sql, Some(service)).map { name =>
           new PreparedStatementImpl(name, service)
+        }.onFailure {
+          err => service.close()
```
Shouldn't' we just close it anyway (no matter if it's a failure or success)?

``` scala
.ensure(service.close())
```

jeremyrsmith replied:
You can't close it at this point if it's a success, because the service still has more phases to execute after Parse.

jeremyrsmith replied:
The `PreparedStatementImpl` needs to use the same service that `Parse` was executed on (because that session is where the prepared statement lives)

jeremyrsmith committed:
Must issue sync as well

Sync needs to be issued on the connection before the service is closed.  Also, prepareAndExecute needs the same treatment.
